### PR TITLE
Adding mask to open_rasterio

### PR DESCRIPTION
**Pull request by jsignell**
- [x] Discussed in #1736
 - [x] Tests added
 - [ ] Fully documented, including `whats-new.rst` for all changes and `api.rst` for new API

Not sure if this is the right approach @snowman2

### CODE REVIEW

**Comment by snowman2** [comment]
@jsignell, this looks like it would work - I like the implementation. Some things I would recommend to add is checks for when the nodata value varies across bands (maybe skip the mask and raise a warning) and when the nodata value is already NaN. Also, you should probably set the nodatavals to NaN if masked to resemble the behavior of xarray elsewhere.

I think @fmaussion would be a good reviewer for this PR.

File: xarray/backends/rasterio_.py [hunk 1]
```diff
@@ -298,6 +300,10 @@ def open_rasterio(filename, parse_coordinates=None, chunks=None, cache=None,
     result = DataArray(data=data, dims=('band', 'y', 'x'),
                        coords=coords, attrs=attrs)
 
+    if mask:
+        for nodataval in attrs.get('nodatavals', ()):
+            result = result.where(result != nodataval)
```
Comment (shoyer) [inline xarray/backends/rasterio_.py]: The downside of this approach is that unlike how to decode metadata for netCDF files, this won't be lazy.

See `CFMaskCoder` in xarray/coding/variables.py for who we do this with netCDF files. In particular, it probably makes sense to use `lazy_elemwise_func` and to move `nodatavals` from `attrs` to `encoding`.

**Comment by fmaussion** [comment]
I wonder if `nodatavals` is the only attribute for masking - in the doc (https://rasterio.readthedocs.io/en/latest/topics/masks.html) they also mention `nodata`, which in my experience is also used

**Comment by snowman2** [comment]
I think the solution would be to add a masked argument and property to the class RasterioArrayWrapper [here](https://github.com/pydata/xarray/blob/778ffc49135d6f97e17b37b48304995fca72f1e0/xarray/backends/rasterio_.py#L24).

And pass in the (mased=masked) argument [here](https://github.com/pydata/xarray/blob/778ffc49135d6f97e17b37b48304995fca72f1e0/xarray/backends/rasterio_.py#L291)

That would pass riods.read( ..., masked=self.masked) when reading data from rasterio [here](https://github.com/pydata/xarray/blob/778ffc49135d6f97e17b37b48304995fca72f1e0/xarray/backends/rasterio_.py#L115)

**Comment by jsignell** [comment]
I was wondering if we should just pass the mask option to rasterio like that. I'll try it out.

**Comment by jsignell** [comment]
I did as @snowman2 suggested but filled the mask with nans at the end. Since it is now using rasterio directly this should address @fmaussion's concern about `nodata` (which I think was ending up in the `nodatavals` attr anyways). I also moved the `nodatavals` from `attrs` to `encoding` when `mask==True`,  but I didn't set it to true by default since I didn't want to break backwards compatibility.

File: xarray/backends/rasterio_.py [hunk 2]
```diff
@@ -268,9 +274,13 @@ def open_rasterio(filename, parse_coordinates=None, chunks=None, cache=None,
         attrs['is_tiled'] = np.uint8(riods.is_tiled)
     if hasattr(riods, 'nodatavals'):
         # The nodata values for the raster bands
-        attrs['nodatavals'] = tuple(
+        nodatavals = tuple(
             np.nan if nodataval is None else nodataval
             for nodataval in riods.nodatavals)
+        if mask:
+            encoding['nodatavals'] = nodatavals
```
Comment (snowman2) [inline xarray/backends/rasterio_.py]: I thing setting the encoding should inline with the other if statements and should be something like:
```python
if mask and riods.nodata is not None:
    encoding['_FillValue'] = riods.nodata
```

Comment (snowman2) [reply xarray/backends/rasterio_.py]: @shoyer is this what you intended or did you want to keep it as @jsignell modified it to be?

**Comment by fmaussion** [comment]
> Do we really want a np.masked_array or do we just want an array with np.Nans in it? I prefer nans, but I guess the downside is type conversion for ints?

Masked arrays come with huge performance drops, I don't think we should use these.

**Comment by shoyer** [comment]
xarray will autoconvert a masked array into a numpy array with NaNs. It
probably doesn't make a difference where this happens.

**Comment by fmaussion** [comment]
I'm sorry this has stalled now - I think this is a nice addition. 

@jsignell : would you have time to finish this up?

**Comment by jsignell** [comment]
Yeah I have some time to finish it up. Looking back at it now it seems like `masked` would be a better kwarg for open_rasterio. Is there anything else that should be changed?

**Comment by snowman2** [comment]
I think it would be good to write the dataset to a netcdf, and read it back in with xarray.open_dataset to ensure everything encodes/decides properly.

**Comment by jsignell** [comment]
> I think it would be good to write the dataset to a netcdf, and read it back in with xarray.open_dataset to ensure everything encodes/decides properly.

Do you mean write a test to do this, or just check it locally?

**Comment by snowman2** [comment]
Good question, I was thinking adding tests.

File: xarray/backends/rasterio_.py [hunk 3]
```diff
@@ -39,6 +39,7 @@ def __init__(self, manager, lock, vrt_params=None):
         if not np.all(np.asarray(dtypes) == dtypes[0]):
             raise ValueError('All bands should have the same dtype')
         self._dtype = np.dtype(dtypes[0])
```
Comment (snowman2) [inline xarray/backends/rasterio_.py]:
```suggestion
        self._dtype = np.float64 if self.masked else np.dtype(dtypes[0])
```

File: xarray/backends/rasterio_.py [hunk 4]
```diff
@@ -119,7 +120,9 @@ def _getitem(self, key):
                 riods = self.manager.acquire(needs_lock=False)
                 if self.vrt_params is not None:
                     riods = WarpedVRT(riods, **self.vrt_params)
-                out = riods.read(band_key, window=window)
+                out = riods.read(band_key, window=window, masked=self.masked)
+                if self.masked:
+                    out = np.ma.filled(out, np.nan)
```
Comment (snowman2) [inline xarray/backends/rasterio_.py]:
```suggestion
                    out = np.ma.filled(out.astype(np.float64), np.nan)
```

File: xarray/backends/rasterio_.py [hunk 5]
```diff
@@ -319,15 +329,15 @@ def open_rasterio(filename, parse_coordinates=None, chunks=None, cache=None,
...
+        RasterioArrayWrapper(manager, lock, vrt_params, masked=masked))
 
     # this lets you write arrays loaded with rasterio
     data = indexing.CopyOnWriteArray(data)
     if cache and chunks is None:
         data = indexing.MemoryCachedArray(data)
 
     result = DataArray(data=data, dims=('band', 'y', 'x'),
-                       coords=coords, attrs=attrs)
+                       coords=coords, attrs=attrs, encoding=encoding)
```
Comment (snowman2) [inline xarray/backends/rasterio_.py]:
```suggestion
                       coords=coords, attrs=attrs)
        result.encoding = encoding
```

**Comment by snowman2** [comment]
@jsignell, the `masked` kwarg is in `rioxarray` if you need it: https://corteva.github.io/rioxarray/html/examples/clip_geom.html

**Comment by keewis** [comment]
this has been unintentionally closed when renaming the branch to merge into to `main`, but given that we consider deprecating `open_rasterio` in favor of the `rasterio` engine provided by `rioxarray` (which seems to already support the `mask` kwarg) that's probably not a mistake. If you disagree feel free to reopen this PR.

**Comment by jsignell** [comment]
I'm fine with this being closed. TBH I had totally forgotten about it :grimacing:

**Pull request closed without merging**